File: npm/create/source/projfiles/recipe.ts
import { MetaData } from "../utils.js";

export const enum RecipeType {
    BEHAVIOR_PACK = "behavior_pack",
    RESOURCE_PACK = "resource_pack",
    ADDON = "addon",
}

export const enum RecipeModuleType {
    RESOURCES = "resources",
    DATA = "data",
    SERVER = "server",
}

export interface RecipeModule {
    type: RecipeModuleType,
    version: string,
    uuid: string,
    [MetaData]: undefined | "scripting",
}

export interface RecipeConfig {
    type: RecipeType,
    artifact: string,
}

export interface RecipeHeader {
    name: string,
    version: string,
    uuid: string | undefined,
    uuids: [string, string] | undefined,
}

export interface Recipe {
    config: RecipeConfig,
    header: RecipeHeader,
    modules: RecipeModule[],
    [MetaData]: {
        scripting: boolean,
    },
}

export function availableModules(recipeType: RecipeType): RecipeModuleType[] {
    switch (recipeType) {
        case RecipeType.BEHAVIOR_PACK:
            return [
                RecipeModuleType.DATA,
                RecipeModuleType.SERVER,
            ];
        case RecipeType.RESOURCE_PACK:
            return [
                RecipeModuleType.RESOURCES,
            ];
        case RecipeType.ADDON:
            return [
                RecipeModuleType.RESOURCES,
                RecipeModuleType.DATA,
                RecipeModuleType.SERVER,
            ];
    }
}

export function displayStrRecipeType(recipeType: RecipeType): string {
    switch (recipeType) {
        case RecipeType.BEHAVIOR_PACK:
            return "Behavior Pack";
        case RecipeType.RESOURCE_PACK:
            return "Resource Pack";
        case RecipeType.ADDON:
            return "Add-On";
    }
}

export function displayStrRecipeMod(moduleType: RecipeModuleType): string {
    switch (moduleType) {
        case RecipeModuleType.RESOURCES:
            return "Resources";
        case RecipeModuleType.DATA:
            return "Data";
        case RecipeModuleType.SERVER:
            return "Server";
    }
}

export function createRecipeConfig(recipeType: RecipeType): RecipeConfig {
    return {
        type: recipeType,
        artifact: "{{short_name}}_{{version}}",
    }
}

export function createRecipeHeader(config: RecipeConfig, name: string): RecipeHeader {
    const isAddon = config.type === RecipeType.ADDON;
    return {
        name,
        version: "0.1.0",
        uuid: isAddon ? undefined : crypto.randomUUID(),
        uuids: isAddon ? [crypto.randomUUID(), crypto.randomUUID()] : undefined,
    }
}

export function createRecipeModule(modType: RecipeModuleType): RecipeModule {
    return {
        type: modType,
        version: "0.1.0",
        uuid: crypto.randomUUID(),
        [MetaData]: modType === RecipeModuleType.SERVER ? "scripting": undefined,
    }
}

export function createRecipe(config: RecipeConfig, header: RecipeHeader, modules: RecipeModule[]): Recipe {
    const scripting = (() => {
        for (const mod of modules) {
            if (mod[MetaData] === "scripting") return true;
        }
        return false;
    })();

    return {
        config,
        header,
        modules,
        [MetaData]: {
            scripting,
        }
    }
}

File: internal/projfiles/recipe.go
package projfiles

import (
	"encoding/json"
	"fmt"
	"io"
	"strings"

	"github.com/mcbundle/mcbundle/internal/jsonst"
)

type Recipe struct {
	Type     RecipeType
	Artifact string

	Name    string
	Version *jsonst.SemVer
	UUID    *jsonst.UUID
	UUIDs   [2]*jsonst.UUID

	Modules []Module

	Authors []string
	License string
}

func (recipe *Recipe) DirName() string {
	return strings.ReplaceAll(recipe.Name, " ", "")
}

func (recipe *Recipe) Load(reader io.Reader) (err error) {
	var decoder = json.NewDecoder(reader)
	var recipeContent recipeJson
	err = decoder.Decode(&recipeContent)
	if err != nil {
		return
	}
	err = validateRecipe(&recipeContent)
	if err != nil {
		return
	}

	recipe.Type = recipeContent.Config.Type

	recipe.Name = recipeContent.Header.Name
	recipe.Version = recipeContent.Header.Version
	recipe.UUID = recipeContent.Header.UUID
	recipe.UUIDs = recipeContent.Header.UUIDs

	recipe.Modules = recipeContent.Modules

	recipe.Authors = recipeContent.Meta.Authors
	recipe.License = recipeContent.Meta.License

	var vars = variables{make(map[string]string)}

	vars.set("name", ssCodeRegExp.ReplaceAllString(recipeContent.Header.Name, ""))
	vars.set("raw_name", recipeContent.Header.Name)
	vars.set("dir_name", recipe.DirName())
	vars.set("short_name",
		strings.ToLower(strings.ReplaceAll(
			ssCodeRegExp.ReplaceAllString(recipeContent.Header.Name, ""),
			" ", "-",
		)),
	)
	vars.set("version", recipeContent.Header.Version.String())

	recipe.Artifact = vars.apply(recipeContent.Config.Artifact)

	return
}

func (recipe *Recipe) PackType() PackType {
	return recipe.Type.PackType()
}

// type Recipe

type Module struct {
	Type    ModuleType     `json:"type"`
	Version *jsonst.SemVer `json:"version"`
	UUID    *jsonst.UUID   `json:"uuid"`
}

func (module *Module) BelongsTo() PackType {
	return module.Type.BelongsTo()
}

// type Module

type ModuleType byte

const (
	ModuleTypeData ModuleType = iota + 1
	ModuleTypeServer
	ModuleTypeResources
)

func (moduleType *ModuleType) UnmarshalJSON(data []byte) (err error) {
	var val string
	err = json.Unmarshal(data, &val)
	if err != nil {
		return
	}

	switch val {
	case "data":
		*moduleType = ModuleTypeData
	case "script":
		fallthrough
	case "server":
		*moduleType = ModuleTypeServer
	case "resources":
		*moduleType = ModuleTypeResources
	default:
		err = fmt.Errorf("unknown module type: \"%s\"", val)
	}

	return
}

func (moduleType ModuleType) String() string {
	switch moduleType {
	case ModuleTypeData:
		return "data"
	case ModuleTypeServer:
		return "server"
	case ModuleTypeResources:
		return "resources"
	}
	return ""
}

func (recipeModule ModuleType) BelongsTo() PackType {
	switch recipeModule {
	case ModuleTypeData:
		fallthrough
	case ModuleTypeServer:
		return PackTypeBehavior
	case ModuleTypeResources:
		return PackTypeResources
	}
	panic("unknown ModuleType value")
}

// type ModuleType

type RecipeType byte

const (
	RecipeTypeBehaviorPack RecipeType = iota + 1
	RecipeTypeResourcePack
	RecipeTypeAddOn
)

func (recipeType *RecipeType) UnmarshalJSON(data []byte) (err error) {
	var val string
	err = json.Unmarshal(data, &val)
	if err != nil {
		return
	}

	switch val {
	case "behavior_pack":
		*recipeType = RecipeTypeBehaviorPack
	case "resource_pack":
		*recipeType = RecipeTypeResourcePack
	case "addon":
		*recipeType = RecipeTypeAddOn
	default:
		err = fmt.Errorf("unknown recipe type: \"%s\"", val)
	}

	return
}

func (recipeType RecipeType) PackType() PackType {
	switch recipeType {
	case RecipeTypeBehaviorPack:
		return PackTypeBehavior
	case RecipeTypeResourcePack:
		return PackTypeResources
	case RecipeTypeAddOn:
		panic("RecipeTypeAddon value cannot have a single PackType")
	}
	panic("unknown RecipeType value")
}

func (recipeType RecipeType) AcceptsModuleType(mT ModuleType) bool {
	if recipeType == RecipeTypeAddOn {
		return ((PackTypeBehavior | PackTypeResources) & mT.BelongsTo()) > 0
	} else {
		return recipeType.PackType() == mT.BelongsTo()
	}
}

func (recipeType RecipeType) AcceptsModule(m *Module) bool {
	return recipeType.AcceptsModuleType(m.Type)
}

// type RecipeType
